src/Bedrock.Framework.Experimental/Transports/WebSockets/WebSocketPayloadEncoder.cs

```diff
@@ -0,0 +1,152 @@
+
+            fixed (byte* dataStartPtr = &MemoryMarshal.GetReference(span))
+            {
+                byte* dataPtr = dataStartPtr;
+                byte* dataEndPtr = dataStartPtr + bytesToRead;
+                byte* maskPtr = (byte*)&maskingKey;
+
+                //If there is at least one int of data available
+                if (dataEndPtr - dataStartPtr >= sizeof(int))
+                {
+                    //Start by aligning to an int boundary, makes aligned loads faster later
+                    while ((ulong)dataPtr % sizeof(int) != 0)
+                    {
+                        Debug.Assert(dataPtr < dataEndPtr);
+
+                        *dataPtr++ ^= maskPtr[_currentMaskIndex];
+                        _currentMaskIndex = (_currentMaskIndex + 1) & 3; //Modulus shortcut for power of 2 values for perf, same as _currentMaskIndex % 4
+                    }
+
+                    //We may have moved forward an uneven number of bytes along the masking key,
+                    //generate a new masking key that has those bytes rotated for whole-int unmasking
+                    int alignedMask = (int)BitOperations.RotateRight((uint)_maskingKey, _currentMaskIndex * 8);
+
+                    //If we found we should use SIMD and there is sufficient data to read
+                    if (useSimd && (dataEndPtr - dataPtr) >= Vector<byte>.Count)
+                    {
+                        //Align by whole ints to full SIMD load boundary to avoid a perf penalty for unaligned loads
+                        while ((ulong)dataPtr % (uint)Vector<byte>.Count != 0)
+                        {
+                            Debug.Assert(dataPtr < dataEndPtr);
+
+                            *(int*)dataPtr ^= alignedMask;
+                            dataPtr += sizeof(int);
+                        }
+
+                        //Unmask full aligned vectors at a time
+                        if (dataEndPtr - dataPtr >= Vector<byte>.Count)
+                        {
+                            Vector<byte> maskVector = Vector.AsVectorByte(new Vector<int>(alignedMask));
+
+                            do
+                            {
+                                *(Vector<byte>*)dataPtr ^= maskVector;
+                                dataPtr += Vector<byte>.Count;
+                            }
+                            while (dataEndPtr - dataPtr >= Vector<byte>.Count);
+                        }
+                    }
+
+                    //Process remaining data (or all, if couldn't use SIMD) one int at a time.
+                    while (dataEndPtr - dataPtr >= sizeof(int))
+                    {
+                        *(int*)dataPtr ^= alignedMask;
+                        dataPtr += sizeof(int);
+                    }
+                }
+
+                //Remaining data less than size of int
+                while (dataPtr != dataEndPtr)
+                {
+                    *dataPtr++ ^= maskPtr[_currentMaskIndex];
```

Comment:
Maybe it's better to assign `_currentMaskIndex` to a local variable, so that the JIT can use registers for this. After the loop write it back to the field.

Reply:
That's a good call.